### ISSUE TITLE
Legion: Hang on Summit

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1309]
I'm trying to run process/node instead of process/gpu for S3D on Summit. It seems to freeze on 4 nodes with 1 process each.

There are stack traces from a debug run with `-ll:force_kthreads` here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/

**Comment by syamajala** [comment]
Latest control_replication: a1cff4d3832

**Comment by lightsighter** [comment]
The stack traces don't tell us much other than that two of the shards are waiting for a trace replay to finish, but what really went wrong is some Realm operation inside the trace probably hung. Try running with `-gex:bindhost 0 -gex:bindcuda 0`. If that doesn't work see if it hangs with either `-lg:inorder` or `-lg:no_physical_tracing` or both of them together. If that doesn't work, run with `-level dma=2` and report the log files from all nodes.

**Comment by syamajala** [comment]
It seems to work with `-gex:bindhost 0 -gex:bindcuda 0`

**Comment by lightsighter** [comment]
Can you try each one individually and see which ones makes it better, or whether they are both required?

**Comment by syamajala** [comment]
Only `-gex:bindcuda 0` is required to make it work.

**Comment by streichler** [comment]
@syamajala can you see if (just - no messing with binding options) `-gex:rmaput 0` changes anything?

**Comment by syamajala** [comment]
It works with just `-gex:rmaput 0` and nothing else.

**Comment by streichler** [comment]
ok, please use that as a workaround for now.  I may need to ask you for some experiments and/or debugger spelunking to narrow down the root cause of the problem.

**Comment by syamajala** [comment]
`-gex:rmaput 0` stops working at 64 nodes. I'll see if I can get some stack traces.

**Comment by streichler** [comment]
Does adding `-gex:obcount 1k` to the `-gex:rmaput 0` make 64 nodes start working?  If so, you're running into #1262 also.  (It'd be a little bit surprising for s3d to be seeing this because of its nearest-neighbor communication pattern, but there might be a broadcast at startup that can trigger it.)

**Comment by lightsighter** [comment]
FWIW, if it is still hanging in the same place as the backtraces above suggest, then we are well into the time stepping because we're doing trace replays and communication is at that point guaranteed to be nearest neighbors. It's unclear though if the new hang is occurring in the same place as the old one though.

**Comment by syamajala** [comment]
It ran for a couple timesteps at 64 nodes before freezing with `-gex:obcount 1000 -gex:rmaput 0` (1k did not work, but 1000 did).

**Comment by streichler** [comment]
sorry about the `1k` part - your use of `1000` is actually what I should have asked for.  However, if it doesn't completely fix the problem, it probably isn't the actual issue, and is just changing the timing around.  Please run with
`-gex:rmaput 0 -gex:obcount 1000 -level dma=2,xplan=1`
and (assuming the hang reproduces with the increased logging) put logs somewhere I can look at them.

**Comment by syamajala** [comment]
I put the logs here on sapling: /scratch2/seshu/s3d_summit_logs

**Comment by streichler** [comment]
> I put the logs here on sapling: /scratch2/seshu/s3d_summit_logs

There are no stuck copies in these logs.  We'll need to look for other symptoms (e.g. via capturing backtraces, and/or trying
runs with the much-more-disruptive `-level dma=2,task=2,event=2,inst=1,barrier=2` logging).

**Comment by syamajala** [comment]
I put some new logs with `-level dma=2,task=2,event=2,inst=1,barrier=2` here: /scratch2/seshu/s3d_summit_logs

**Comment by lightsighter** [comment]
These logs are from a run that hung?

**Comment by syamajala** [comment]
I believe so. It never completed the first timestep and the job was killed after 30 minutes.

**Comment by streichler** [comment]
There are suspended Legion metatasks on at least ranks 0 and 1 (the only two I checked) in the most recent logs.  Backtraces (with `-ll:force_kthreads`) are probably the next step?

**Comment by lightsighter** [comment]
I'm pretty suspicious that Legion is going to be the cause here given the prior evidence that the hang goes away with altered GASNet settings, and we already know that there are non-trivial  but also non-interesting backtraces for the first hang in this issue. What do we want to learn from the backtraces?

**Comment by streichler** [comment]
Unless I've lost track of things, this hang is happening even with the altered GASNet settings, and as a result I believe it to be a distinct issue from the rmaput and gex obcount bugs.  @syamajala can you either confirm that or correct me?

**Comment by syamajala** [comment]
It works with altered gasnet settings up to 64 nodes and then begins to hang again. 

The flags I tried were: `-gex:bindhost 0 -gex:bindcuda 0` (works), `-gex:bindcuda 0` (works), and `-gex:rmaput 0` (works).

**Comment by lightsighter** [comment]
Right, so we can get the new backtraces, but I'm skeptical that they're going to tell us anything interesting.

**Comment by syamajala** [comment]
I got some stacktraces with `-gex:rmaput 0 -gex:obcount 1000 -ll:force_kthreads 1`.

They are available here on sapling: `/scratch2/seshu/s3d_summit_logs/stacktraces`

**Comment by lightsighter** [comment]
We're still capturing the trace. Are you sure this is actually a hang and not just taking a long time to capture the trace? It would be good to check that the stack traces are not changing over time. Ones like this show we're still capturing the trace, but if they don't change then that means that there are outstanding message tasks (with no event preconditions) that are not running:

```
Thread 71 (Thread 0x201d405ef890 (LWP 70943)):
#0  0x0000200000a25e84 in syscall () from /lib64/power9/libc.so.6
#1  0x0000200003b44d34 in Realm::Doorbell::wait_slow (this=0x201d405f0158) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/r
untime/realm/mutex.cc:316
#2  0x0000200003b466c8 in Realm::Doorbell::wait (this=0x201d405f0158) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtim
e/realm/mutex.inl:81
#3  Realm::FIFOCondVar::wait (this=0x201d405ed1f0) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/realm/mutex.cc:10
88
#4  0x0000200003aa02f4 in Realm::KernelThreadTaskScheduler::worker_sleep (this=0x2533aba0, switch_to=0x2009700b8710) at /gpfs/alpine/cmb103/scratc
h/seshuy/legion_s3d_nscbc_startup/legion/runtime/realm/tasks.cc:1414
#5  0x0000200003aa2414 in Realm::ThreadedTaskScheduler::thread_blocking (this=0x2533aba0, thread=<optimized out>) at /gpfs/alpine/cmb103/scratch/s
eshuy/legion_s3d_nscbc_startup/legion/runtime/realm/tasks.cc:908
#6  0x0000200003990a5c in Realm::Thread::wait_for_condition<Realm::EventTriggeredCondition> (cond=..., poisoned=@0x201d405edae8: false) at /gpfs/a
lpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/realm/event_impl.cc:160
#7  0x000020000397b780 in Realm::Event::wait_faultaware (this=0x201d405edbc8, poisoned=@0x201d405edae8: false) at /gpfs/alpine/cmb103/scratch/sesh
uy/legion_s3d_nscbc_startup/legion/runtime/realm/event_impl.cc:254
#8  0x000020000397bf7c in Realm::Event::wait (this=<optimized out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/
realm/event_impl.cc:206
#9  0x0000200001d709f4 in Legion::Internal::LgEvent::wait (this=<optimized out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/le
gion/runtime/legion/legion_types.h:2833
#10 0x000020000240546c in Legion::Internal::RemoteTraceRecorder::record_create_ap_user_event (this=0x200970509710, lhs=..., tlid=...) at /gpfs/alp
ine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/legion_analysis.cc:457
#11 0x0000200002423668 in Legion::Internal::TraceInfo::record_create_ap_user_event (result=..., this=<optimized out>) at /gpfs/alpine/cmb103/scrat
ch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/legion_analysis.h:494
#12 Legion::Internal::Runtime::create_ap_user_event (info=0x20097089c260) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/ru
ntime/legion/runtime.h:5112
#13 Legion::Internal::UpdateAnalysis::perform_remote (this=0x20097089c090, perform_precondition=..., applied_events=..., already_deferred=<optimiz
ed out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/legion_analysis.cc:6663
#14 0x000020000242d624 in Legion::Internal::UpdateAnalysis::handle_remote_updates (derez=..., runtime=<optimized out>, previous=<optimized out>) a
t /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/legion_analysis.cc:6930
#15 0x00002000028f26a4 in Legion::Internal::Runtime::handle_equivalence_set_remote_updates (this=<optimized out>, derez=..., source=<optimized out
>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/runtime.cc:24905
#16 0x00002000029b6e44 in Legion::Internal::VirtualChannel::handle_messages (this=0x201c5007cd00, num_messages=<optimized out>, runtime=0x2ef2dcf0
, remote_address_space=<optimized out>, args=<optimized out>, arglen=<optimized out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_start
up/legion/runtime/legion/runtime.cc:12613
#17 0x00002000029b766c in Legion::Internal::VirtualChannel::process_message (this=0x201c5007cd00, args=<optimized out>, arglen=9976, runtime=0x2ef
2dcf0, remote_address_space=<optimized out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/runtime.cc:11602
#18 0x00002000029b79e8 in Legion::Internal::MessageManager::receive_message (this=<optimized out>, args=<error reading variable: value has been op
timized out>, arglen=<optimized out>) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/runtime.cc:13065
#19 0x00002000029b7a4c in Legion::Internal::Runtime::process_message_task (this=<optimized out>, args=0x201dff36110c, arglen=<optimized out>) at /
gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/runtime.cc:25887
#20 0x00002000029b7d20 in Legion::Internal::Runtime::legion_runtime_task (args=0x201dff361100, arglen=9992, userdata=<optimized out>, userlen=<opt
imized out>, p=...) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/legion/runtime.cc:31383
#21 0x0000200003a46620 in Realm::LocalTaskProcessor::execute_task (this=0x2e742010, func_id=<optimized out>, task_args=...) at /gpfs/alpine/cmb103
/scratch/seshuy/legion_s3d_nscbc_startup/legion/runtime/realm/bytearray.inl:150
#22 0x0000200003aa0720 in Realm::Task::execute_on_processor (this=0x201dff32bc90, p=...) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_st
artup/legion/runtime/realm/runtime_impl.h:395
#23 0x0000200003aa0894 in Realm::KernelThreadTaskScheduler::execute_task (this=<optimized out>, task=<optimized out>) at /gpfs/alpine/cmb103/scrat
ch/seshuy/legion_s3d_nscbc_startup/legion/runtime/realm/tasks.cc:1366
#24 0x0000200003aa38d4 in Realm::ThreadedTaskScheduler::scheduler_loop (this=0x2533aba0) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_st
artup/legion/runtime/realm/tasks.cc:1103
#25 0x0000200003aa40dc in Realm::ThreadedTaskScheduler::scheduler_loop_wlock (this=0x2533aba0) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_ns
cbc_startup/legion/runtime/realm/tasks.cc:1217
#26 0x0000200003aa7720 in Realm::KernelThread::pthread_entry (data=0x201d8c0e6220) at /gpfs/alpine/cmb103/scratch/seshuy/legion_s3d_nscbc_startup/
legion/runtime/realm/threads.cc:774
#27 0x00002000008b8ae0 in start_thread () from /lib64/power9/libpthread.so.0
#28 0x0000200000a2e7c8 in clone () from /lib64/power9/libc.so.6
```

**Comment by syamajala** [comment]
It seems to be very slow. It took about 10 minutes to finish first time step on 4 nodes, we're running rank/node too. Maybe I'll try turning tracing off?

**Comment by lightsighter** [comment]
Why don't you try capturing a profile first so we can actually understand what is going on.

**Comment by syamajala** [comment]
Here is a profile from the first timestep on rank 0: http://sapling.stanford.edu/~seshu/s3d_summit_hang/legion_prof/?start=0&end=531388052.90000004&collapseAll=false&resolution=10

I'm not able to open the profile when I add other ranks to it. I'll generate a few other ranks separately.

**Comment by lightsighter** [comment]
It looks to me like you have a region without a used disjoint and complete partition, although it would be good to see the profiles for the other nodes too (you can make a separate profile for each of the four nodes). You do have to guarantee that every region at the moment has at least one disjoint and complete partition that you've used for scalability reasons right now. That is one of the reasons that `control_replication` has not merged into master yet is that needs to be fixed.

**Comment by syamajala** [comment]
I ran with `-lg:partcheck 1` on 16 nodes and it passed. Every partition is made with either`restrict` (and explicitly marked as `disjoint` and `complete`) or `equal`. 

One thing though is that we have some tasks that are launched over a mix of rank and subrank level partitions.

**Comment by lightsighter** [comment]
Do all the profiles look like this? There are just huge amounts of time where nothing is happening anywhere. If it is the equivalence set problem, then there is just going to be one or a few nodes where everything is happening, but if it is that then you should be getting this warning:
https://gitlab.com/StanfordLegion/legion/-/blob/control_replication/runtime/legion/legion_analysis.cc#L10320-10325

> I ran with -lg:partcheck 1 on 16 nodes and it passed. Every partition is made with eitherrestrict (and explicitly marked as disjoint and complete) or equal.

And these partitions are being used and not just created?

> Is there a way to tell what region the runtime thinks does not have a disjoint and complete partition?

Yes, you'll see the warning above.

**Comment by lightsighter** [comment]
The other thing which could be going wrong is you just have the sharding very wrong and lots of communication is required as a result.

**Comment by syamajala** [comment]
I generated profiles for every rank: http://sapling.stanford.edu/~seshu/s3d_summit_hang/

I looked through a few and they seem similar to ranks 31 and 62 to me. 

I never saw that warning for the equivalent sets in any of the logs. 

I think some of the rank level partitions are just created and never used, but all the subrank partitions should be used. 

I am using the default sharding functor. That made the most sense to me for rank/node. Is there a better one I should be using?

**Comment by lightsighter** [comment]
You're sure you are not suppressing the warnings with compile time logging settings or command line flags?

**Comment by syamajala** [comment]
I am seeing warnings just not the one related to equivalence sets. 

This is one I see: 
```
[0 - 155510591840]   13.827477 {4}{runtime}: [warning 1107] LEGION WARNING: We strongly encourage all projection functors to be functional, however, projection function 58 is not and therefore an expensive analysis cannot be memoized. Please consider making it functional to avoid performance degredation. (from file /lustre/scratch/vsyamaj/legion_s3d_test/legion/runtime/legion/legion_analysis.cc:3166)
```

But @elliottslaughter told me the projection functor warning would go away with tracing.

**Comment by syamajala** [comment]
I used spy and checked the region graph, which is available here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/region_graph.pdf

Every region has a disjoint and complete partition. The only partitions which are disjoint but not complete are the ones for ghost regions for the stencils.

**Comment by lightsighter** [comment]
Which commit caused this behavior to change?

**Comment by syamajala** [comment]
What behavior? I don't think anything has changed?

**Comment by lightsighter** [comment]
I'm assuming this was working on Summit before and now you're saying that something has changed to produce this behavior.

**Comment by syamajala** [comment]
This is my first time trying to scale the subrank branch of S3D on Summit.

**Comment by syamajala** [comment]
I generated the profiles with the rust profiler. I will try the python profiler and see if it gives any messages. 

Everything should be sharded. Looking at the mapper I don't think its possible to do remote mappings? It does a query of local processors when constructing the mapper and saves those to vectors then just uses those when figuring out the mappings.

**Comment by syamajala** [comment]
This message actually shows up at 2 nodes. 

```
WARNING: A significant number of long latency messages were detected during this run meaning that the network was likely congested and could be causing a significant performance degredation. We detected 2863 messages that took longer than 1000.00us to run, representing 18.57% of 15414 total messages. The longest latency message required 33617.80us to execute. Please report this case to the Legion developers along with an accompanying Legion Prof profile so we can better understand why the network is so congested.
```

**Comment by syamajala** [comment]
Using the logging mapper I confirmed there are no remote mappings.

**Comment by lightsighter** [comment]
Can you try running after building with `-DDISABLE_EQUIVALENCE_SET_MIGRATION`?

**Comment by syamajala** [comment]
Here is what I see when `-DDISABLE_EQUIVALENCE_SET_MIGRATION`: http://sapling.stanford.edu/~seshu/s3d_summit_hang/legion_prof_64ranks_no_equiv_migration

Took about 1 minute for startup. I submitted a longer run as well.

**Comment by lightsighter** [comment]
In general, the smaller you make this example in code or number of nodes and still see the behavior the easier it will be to debug.

**Comment by syamajala** [comment]
Here is the startup time from 1 to 64 nodes with equivalence set migration: http://sapling.stanford.edu/~seshu/s3d_summit_hang/migration/startup_legion_prof.html
Profiles: http://sapling.stanford.edu/~seshu/s3d_summit_hang/migration/

Here is startup time from 1 to 64 nodes without equivalence set migration:
http://sapling.stanford.edu/~seshu/s3d_summit_hang/no_migration/startup_legion_prof.html
Profiles:  http://sapling.stanford.edu/~seshu/s3d_summit_hang/no_migration/

For comparison the startup time when running rank/gpu at 128 nodes (768 ranks in total) is 40 seconds. So 1 minute at 64 ranks is not great.

**Comment by syamajala** [comment]
I ran 6 ranks/node using 1, 2, 4, and 6 cpus: http://sapling.stanford.edu/~seshu/s3d_summit_hang/6rank/startup_legion_prof.html

**Comment by lightsighter** [comment]
Can you make a four-node six-cpu-per-node configuration for me to run on sapling with the runtime compiled in debug mode?

**Comment by syamajala** [comment]
I had to do 1 node with 4 ranks using 6 cpus/rank. If you really want 4 nodes 6 cpus/node I'll ask people for to give me the nodes on sapling. 

There is a script here: `/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases/subrank.sh` 

Just get an interactive job and do ./subrank.sh in /scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases

**Comment by lightsighter** [comment]
Can you try with this branch?

https://gitlab.com/StanfordLegion/legion/-/merge_requests/560

**Comment by syamajala** [comment]
Looks like there is some improvement.

Before: http://sapling.stanford.edu/~seshu/s3d_summit_hang/6rank/6cpu/startup_legion_prof.html
After: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/startup_legion_prof.html

**Comment by lightsighter** [comment]
It's not that much better. The line should be flat. Please make more profiles. They don't need to be at the largest scale, just at some scale where the effect starts to occur.

**Comment by lightsighter** [comment]
Also make region tree pictures which should now show which node is owning each of the nodes.

**Comment by syamajala** [comment]
Profiles from 4 - 32 nodes are available here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/

I will make a picture of the region tree.

**Comment by syamajala** [comment]
There is a region tree from a 4 node/24 rank run here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/tmp24_0/region_graph.pdf

Dot file too: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/tmp24_0/region_graph.dot

If you want a region tree from a bigger run I can try to generate one.

**Comment by lightsighter** [comment]
We really want the index space trees and preferably for a region tree that can actually be visualized.

**Comment by syamajala** [comment]
I did some smaller runs on sapling: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/spy/

The shardrefine branch seems to be a bit unstable. Some configurations I tried on sapling seem to freeze at start up so I can't go much smaller right now. 

There is are two processes on c0004, pids 40425 and 40426

**Comment by syamajala** [comment]
Looks like the smaller run actually took 30 minutes to startup on sapling. It was 2 ranks, 4 cpus/rank.

**Comment by lightsighter** [comment]
I really doubt that this code is hanging.

I'm still having trouble understanding these region trees. They don't align with anything that I would expect for S3D. There's lots of what appear to be symmetric partitions of the same regions. I haven't checked them all by hand though to ensure that they are all really congruent or not. It makes it hard to know which partitions are actually being used and which ones are just there.

Make all the trees and a profile from the same run on sapling. Use `-lg:footprint 0` so you can kill it after 10-15 minutes of execution and we'll actually be able to look at what is going on in the profile.

If you have spare time you can also make me a reproducer, but do everything else first.

**Comment by syamajala** [comment]
I tried `-lg:footprint 0` and killed it after 20 minutes, but the profiler logs are always empty. I think this is hanging. 

I left two processes running on c0003. PIDs 232496 and 232497.

**Comment by syamajala** [comment]
Actually looks like its hanging on the fortran side sometimes. I turned off mpiio and it seems to be ok now.

**Comment by syamajala** [comment]
There is a profile here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/sapling/legion_prof/

Spy graphs here: http://sapling.stanford.edu/~seshu/s3d_summit_hang/shardrefine/sapling/spy/

**Comment by lightsighter** [comment]
Which branch is this profile from?

**Comment by syamajala** [comment]
That was with the shardrefine branch.

**Comment by lightsighter** [comment]
This profile looks fine to me. Utility work is balanced across both the threads. It takes about 8 seconds to do the physical analysis for the first iteration and capture the trace. There three seconds for the transitive reduction, but replays can start even before that finishes so it'll just be running in the background.

Is there something I'm missing that you think is wrong?

**Comment by syamajala** [comment]
The behavior on sapling has been very inconsistent for some reason. I tried some runs on perlmutter. 

Here is rank/node with the shardrefine branch: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/startup_legion_prof.html
Profiles: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine

Here is rank/node with control_replication: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/ctrlrepl/startup_legion_prof.html
Profiles: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/ctrlrepl/

Past 64 nodes with both control_replication and shardrefine we're hitting oom.

Here is rank/gpu with control_replication for comparison: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/rank_per_gpu/startup_legion_prof.html
Profiles: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/rank_per_gpu/

When running rank/gpu it takes 64 seconds for 1024 ranks to startup vs 97 seconds for 64 ranks with shardrefine and 81 seconds for 64 ranks with control_replication.

**Comment by lightsighter** [comment]
Rerun the 8, 16, and 32 rank profiles after building with `-DDISABLE_EQUIVALENCE_SET_MIGRATION`. From now on, just leave that on permanently unless I tell you otherwise. We shouldn't need to be doing equivalence set migration at all for S3D and yet it is dominating all your performance which means something else is wrong, but it will be impossible to debug if the runtime is trying to work around the problem with very slow equivalence set migration.

The 64 rank profiles are practically unusable. Whatever is going on there will be obviously visible at smaller scales too. Do not waste your time and hours on the bigger node counts until we have confirmed that the profiles at smaller node counts look impeccable and we can explain every single performance phenomenon that we observe there.

> The behavior on sapling has been very inconsistent for some reason.

Unless you want to narrow down the parts of S3D into a much smaller program that exhibits the same bad equivalence set behavior then I strongly suggest that you find a way to reproduce this on sapling so I can run it for myself.

**Comment by syamajala** [comment]
Here are profiles for the new runs: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/

Things looked ok to me up to 32 so I went a bit further: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/startup_legion_prof.html

Compared to rank/gpu: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/startup_legion_prof.html

I am looking at reproducing on sapling.

**Comment by lightsighter** [comment]
Ok, so now that the equivalence set migration isn't muddying the waters anymore, the pattern starts to emerge. It looks like all the equivalence sets for the stencils are being sharded onto node 0 as we see giant numbers of physical analyses showing up on node 0's utility processor, and hardly any remote analyses showing up on other utility processors. This is almost certainly going to be a function of your region tree and which node made the partitions. I can't see any patterns from the index space trees above since they look like they are only from a 2 node run. Let's see what an index space tree from a 4 node runs looks like.

**Comment by syamajala** [comment]
The index space tree from a 4 node run is available here: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/pwave_x_4_hept/

1 and 2 nodes are also available: http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/pwave_x_1_hept/

http://sapling.stanford.edu/~seshu/s3d_perlmutter_subrank/shardrefine/pwave_x_2_hept/

Perlmutter is down for the rest of the week.

**Comment by lightsighter** [comment]
These partitions still don't align with what I would expect for S3D. Which ones are being used for actually performing the computations?

**Comment by syamajala** [comment]
There is a job script here: `/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases/subrank.sh`

Its setup for 2 nodes (c0003, c0004), 4 ranks/node, 6 subranks/rank. Built in debug with spy.

**Comment by lightsighter** [comment]
I'm pretending to be you when launching the job, but still getting an error about $LD_LIBRARY_PATH being unbound from slurm:

```
seshu@sapling:/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases$ sbatch subrank.sh
Submitted batch job 15353
seshu@sapling:/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases$ cat slurm-15353.out 
environment: /usr/local/modules-3.2.10/bin/modulecmd: No such file or directory
/var/spool/slurm/slurmd/job15353/slurm_script: line 15: LD_LIBRARY_PATH: unbound variable
```

If I comment out the depending on a prior definition of the `$LD_LIBRARY_PATH` in the script, then I get errors about mpirun not being available in your environment:

```
environment: /usr/local/modules-3.2.10/bin/modulecmd: No such file or directory
{'-x': '../fortran/hept/s3d.x', '-l': '../build/hept/librhst.so', '-e': 'RHST_ARGS=-ll:cpu 6 -ll:util 1 -ll:bgwork 1 -ll:csize 16384 -ll:rsize 1024 -ll:gsize 0 -hl:safe_mapper -ll:stacksize 16 -logfile run_%.log -lg:prof 24 -lg:prof_logfile prof_%.gz -lg:spy 1 -lg:delay 20', '-g': '32x32x96', '-v': '1x1x1', '-t': '1', '-b': '../s3d/', '-p': '2x2x2', '-s': 'mpi', '-H': 'c0001,c0003', '--bomb-file': '0d', '-k': './pwave_x_2_hept', '-o': './pwave_x_2_hept/ammonia.txt', '--title': 'pressure_wave_test'}
Mechanism appears to be: C7H16_52species
copying following file: ../build/hept/librhst.so
to tlib: /scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases/pwave_x_2_hept/librhsf.so
/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases/pwave_x_2_hept:/scratch2/seshu/legion_s3d_no_cuda//build/hept:/scratch2/seshu/legion_s3d_no_cuda//kernels/hept_kernels:/scratch2/seshu/legion_s3d_no_cuda//legion/language/build/lib
['mpirun', '-n', '8', '-N', '4', '-H', 'c0001,c0003', '--report-bindings', '-x', 'LD_LIBRARY_PATH', '--bind-to', 'none', '--oversubscribe', '-x', 'RHSF_MAPMODE=allcpu', '-x', 'REALM_FREEZE_ON_ERROR=1', '-x', 'REALM_BACKTRACE=0', '-x', 'RHSF_FUSE=200', '-x', 'RHST_TRACE=1', '-x', 'RHST_PERFORM_DUMPS=0', '-x', 'RHST_ARGS=-ll:cpu 6 -ll:util 1 -ll:bgwork 1 -ll:csize 16384 -ll:rsize 1024 -ll:gsize 0 -hl:safe_mapper -ll:stacksize 16 -logfile run_%.log -lg:prof 24 -lg:prof_logfile prof_%.gz -lg:spy 1 -lg:delay 20', '/scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases/pwave_x_2_hept/s3d.x']
Traceback (most recent call last):
  File "./subrank.py", line 557, in <module>
    run_s3d(opts, envvars)
  File "./subrank.py", line 351, in run_s3d
    subprocess.call(cmdline, cwd = rdir, stdin = None, stdout = outf, stderr = outf)
  File "/usr/lib/python3.8/subprocess.py", line 340, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'mpirun'
```

**Comment by syamajala** [comment]
I just ssh into c0003 and run the script directly:
```
cd /scratch2/seshu/legion_s3d_no_cuda/Ammonia_Cases
./subrank.sh
```

**Comment by syamajala** [comment]
Since we're running on sapling, would it be worth getting vtune profiles?

**Comment by lightsighter** [comment]
Everything looks ok at first pass on this code. I've modified your runtime. Can you rebuild it and let me know when it is ready to test? We'll see if we hit the interesting assertion I put it to test that nothing other than stencil tasks are performing remote analysis.

> Since we're running on sapling, would it be worth getting vtune profiles?

Not for fixing this particular problem. I would be curious to see if there are any hotspots in the runtime to performance tune if you want to do that, but vtunes won't tell us what is going wrong here.

**Comment by syamajala** [comment]
I rebuilt the runtime and ran it. Seem to hit the new assertion you added.

**Comment by syamajala** [comment]
Yes, that is the whole simulation bound.

**Comment by syamajala** [comment]
Is there anything new for me to try for this issue? I looked through S3D and we are never passing the top level regions directly to any tasks. The mapper also does not allow us to map top level regions.

**Comment by lightsighter** [comment]
I don't think there's anything to do at the moment. I need to dig into the heuristic that decides how to refine equivalence sets because it's clearly not making refinements in some places where it should. Unfortunately that is some detailed code that has long fallen out of my head and need a dedicated time to do it. I was planning to work on it this weekend, but some Legate features took priority. I'm currently slating it for this weekend.

**Comment by lightsighter** [comment]
I did not get to this during the weekend. PPoPP rebuttal and too many non-Legion distractions. It's still near the top of my non-NVIDIA priorities.

**Comment by lightsighter** [comment]
@syamajala Did you rebuild this? I'm not seeing the error anymore for the assertion that I added.

**Comment by syamajala** [comment]
I rebuilt and hit the assertion. Should be good for you to test.

**Comment by lightsighter** [comment]
I at least understand what is causing the lack of refinement. This shouldn't matter, but at the moment it does: right now your non-identity (e.g. projection function id > 0) projection functions are marked as "not functional" and that's preventing the current heuristics from performing the needed refinements. If you can mark them as functional, that will likely help things. I'm developing some better heuristics now and this case will get fixed as well as part of that process.

**Comment by syamajala** [comment]
I remember talking to @elliottslaughter about this a while ago. There is a reason why Regent does not currently mark projection functors as functional, but he will to comment on why that is.

**Comment by elliottslaughter** [comment]
This is a multi-level projection functor. Basically it looks something like:

```
for point in ... do
  my_task(p[point.x][point.y])
end
```

For a one-level projection functor, the partition `p` is fully captured and described by the value passed into the region requirement. But for a two-level projection functor, this is only the upper bound: in addition to the first-level partition, we also need to know the color of the second-level partition to look up. Regent generates a closure for the projection functor and that closure captures data, making it impossible (if I recall correctly) to use the functional API for Legion projection functors at all.

In other words, it is not merely a matter of setting a boolean flag that says, "it's ok, this is pure". Legion functional and non-functional projection functors expose entirely different APIs that provide different capabilities, and what we need doesn't fit within the scope of the functional API.

**Comment by lightsighter** [comment]
Yup, I understand. I already mentioned that I need to fix this in Legion anyway, but it would at least unblock @syamajala if it was easy to change.

**Comment by syamajala** [comment]
Just wondering if there is a guess as to when this might be available?

**Comment by lightsighter** [comment]
If you're referring to new refinement heuristics, the collective branch will merge this week, and then I would hope to fix the refinement algorithm in the next two weeks after that. @elliottslaughter will need to comment on the projection part.

**Comment by elliottslaughter** [comment]
What I said in https://github.com/StanfordLegion/legion/issues/1309#issuecomment-1304714999 still holds. The projection functors are not going to be marked functional because there is no way for the compiler to do so.

**Comment by syamajala** [comment]
I looked at `advise_analysis_subtree` and it wasn't clear to me that it would help with this specific issue? Is there a chance we can get some short term hack for directing the placement of equivalence sets? 

Due to the libfabric issues on Frontier, subranks for S3D is our only hope left to do Gordon Bell.

**Comment by lightsighter** [comment]
> Is there a chance we can get some short term hack for directing the placement of equivalence sets?

That's exactly what `advise_analysis_subtree` does: it tells the runtime which regions and partitions to use for equivalence sets.

**Comment by syamajala** [comment]
I tried this out. It never gets through start up at 4 nodes. Its possible I picked the wrong region, partition, and fields. I'm using the lr_int region, at the rank level, and passing all fields: https://gitlab.com/legion_s3d/legion_s3d/-/blob/subranks/rhst/s3d_rank_wrapper.cc#L703-734

**Comment by lightsighter** [comment]
Make sure your partition is disjoint and complete, and you have to do it for every single field that you care about. You also need to do it separately for every single region tree.

**Comment by syamajala** [comment]
I added all regions and all fields, but it made no difference.

It seems to start up when I use `-lg:inorder` though.

**Comment by lightsighter** [comment]
Are you sure the case without `-lg:inorder` is not just hanging normally? The last time we looked at it we concluded that was not the case, but it might be the case here. The logic that determines how to do the refinements is in the dependence analysis stage of the pipeline and is therefore invariant to any reordering done the mapping stage of the pipline. It will refine the tree the same way regardless of whether you run with `-lg:inorder` or not.

**Comment by syamajala** [comment]
Shardrefine has fixed these problems. We still have issues on Summit, but those are being worked through in #1594. So I will close this.
